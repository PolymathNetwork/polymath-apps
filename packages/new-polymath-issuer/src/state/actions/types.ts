import { ActionType } from 'typesafe-actions';
import * as sessionActions from './session';
import * as appActions from './app';

export type SessionActions = ActionType<typeof sessionActions>;
export type AppActions = ActionType<typeof appActions>;
