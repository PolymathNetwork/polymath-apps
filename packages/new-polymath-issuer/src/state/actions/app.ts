import { createStandardAction } from 'typesafe-actions';

export const setActiveTransactionGroup = createStandardAction(
  'APP/SET_ACTIVE_TRANSACTION_GROUP'
)<string>();

export const initializePolyClientStart = createStandardAction(
  'APP/INITIALIZE_POLY_CLIENT_START'
)();

export const initializePolyClientSuccess = createStandardAction(
  'APP/INITIALIZE_POLY_CLIENT_SUCCESS'
)();

export const initializePolyClientFailure = createStandardAction(
  'APP/INITIALIZE_POLY_CLIENT_FAILURE'
)<Error>();
