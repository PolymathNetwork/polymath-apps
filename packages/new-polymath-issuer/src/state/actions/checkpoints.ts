import { types } from '@polymathnetwork/new-shared';
import { createEntityActions } from '~/state/helpers/createEntityActions';

const { createAction, updateAction, deleteAction } = createEntityActions<
  types.CheckpointEntity
>('CHECKPOINTS');

export { createAction, updateAction, deleteAction };
