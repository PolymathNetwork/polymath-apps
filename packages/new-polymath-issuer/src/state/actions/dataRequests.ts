import { createStandardAction } from 'typesafe-actions';
import { RequestKeys, Fetcher, RequestArgs } from '~/types';

const invalidateRequest = createStandardAction('DATA_REQUESTS/INVALIDATE')<{
  requestKey: RequestKeys;
  args?: RequestArgs;
}>();

const cacheData = createStandardAction('DATA_REQUESTS/CACHE')<{
  requestKey: RequestKeys;
  args: RequestArgs;
  fetchedIds: string[];
}>();

/**
 * This action signals the saga in charge of fetching data from the blockchain
 * that a component needs certain data
 */
const requestData = createStandardAction('DATA_REQUESTS/REQUEST')<{
  fetcher: Fetcher<RequestArgs>;
}>();

const fetchDataStart = createStandardAction('DATA_REQUESTS/FETCH_START')<{
  requestKey: RequestKeys;
  args: RequestArgs;
}>();

const fetchDataFail = createStandardAction('DATA_REQUESTS/FETCH_FAIL')<{
  errorMessage: string;
  requestKey: RequestKeys;
  args: RequestArgs;
}>();

export {
  invalidateRequest,
  cacheData,
  requestData,
  fetchDataStart,
  fetchDataFail,
};
