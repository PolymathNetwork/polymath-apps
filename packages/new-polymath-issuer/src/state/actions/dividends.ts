import { types } from '@polymathnetwork/new-shared';
import { createEntityActions } from '~/state/helpers/createEntityActions';

const { createAction, updateAction, deleteAction } = createEntityActions<
  types.DividendEntity
>('DIVIDENDS');

export { createAction, updateAction, deleteAction };
