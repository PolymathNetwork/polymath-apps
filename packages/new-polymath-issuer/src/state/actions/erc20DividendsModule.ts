import { Erc20DividendsModuleEntity } from '~/types';
import { createEntityActions } from '~/state/helpers/createEntityActions';
import { createStandardAction } from 'typesafe-actions';

const { createAction, updateAction, deleteAction } = createEntityActions<
  Erc20DividendsModuleEntity
>('ERC20_DIVIDENDS_MODULES');

const enableModule = createStandardAction('ERC20_DIVIDENDS_MODULES/ENABLE')<{
  securityTokenSymbol: string;
}>();

export { createAction, updateAction, deleteAction, enableModule };
