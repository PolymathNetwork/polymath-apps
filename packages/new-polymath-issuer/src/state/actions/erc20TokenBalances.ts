import { types } from '@polymathnetwork/new-shared';
import { createEntityActions } from '~/state/helpers/createEntityActions';

const { createAction, updateAction, deleteAction } = createEntityActions<
  types.Erc20TokenBalanceEntity
>('ERC20_TOKEN_BALANCES');

export { createAction, updateAction, deleteAction };
