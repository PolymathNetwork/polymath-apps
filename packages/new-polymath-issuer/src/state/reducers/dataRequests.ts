import { Reducer } from 'redux';
import { getType } from 'typesafe-actions';
import * as actions from '~/state/actions/dataRequests';
import { DataRequestsActions } from '~/state/actions/types';
import { RequestKeys } from '~/types';
import { utils } from '@polymathnetwork/new-shared';

export interface DataRequestResults {
  [argsHash: string]:
    | {
        fetching: boolean;
        fetchedIds: string[] | undefined;
      }
    | undefined;
}

export interface DataRequestsState {
  [RequestKeys.GetCheckpointsBySymbol]: DataRequestResults;
  [RequestKeys.GetCheckpointBySymbolAndId]: DataRequestResults;
  [RequestKeys.GetSecurityTokenBySymbol]: DataRequestResults;
  [RequestKeys.GetDividendsByCheckpoint]: DataRequestResults;
  [RequestKeys.GetErc20DividendsModuleBySymbol]: DataRequestResults;
  [RequestKeys.GetTaxWithholdingListBySymbol]: DataRequestResults;
}

const initialState: DataRequestsState = {
  [RequestKeys.GetCheckpointsBySymbol]: {},
  [RequestKeys.GetCheckpointBySymbolAndId]: {},
  [RequestKeys.GetSecurityTokenBySymbol]: {},
  [RequestKeys.GetDividendsByCheckpoint]: {},
  [RequestKeys.GetErc20DividendsModuleBySymbol]: {},
  [RequestKeys.GetTaxWithholdingListBySymbol]: {},
};

export const reducer: Reducer<DataRequestsState, DataRequestsActions> = (
  state = initialState,
  action
): DataRequestsState => {
  switch (action.type) {
    case getType(actions.invalidateRequest): {
      const {
        payload: { requestKey, args },
      } = action;

      if (args) {
        const argsHash = utils.hashObj(args);
        const {
          [requestKey]: { [argsHash]: invalidData, ...validData },
          ...rest
        } = state;

        return {
          [requestKey]: {
            ...validData,
          },
          ...rest,
        } as DataRequestsState;
      }

      const { [requestKey]: invalidRequest, ...validRequests } = state;

      return {
        [requestKey]: {},
        ...validRequests,
      } as DataRequestsState;
    }
    case getType(actions.cacheData): {
      const {
        payload: { requestKey, args, fetchedIds },
      } = action;

      const argsHash = utils.hashObj(args);

      const {
        [requestKey]: { [argsHash]: invalidData, ...validData },
        ...validRequests
      } = state;

      return {
        [requestKey]: {
          [argsHash]: {
            fetching: false,
            fetchedIds,
          },
          ...validData,
        },
        ...validRequests,
      } as DataRequestsState;
    }
    case getType(actions.fetchDataStart): {
      const {
        payload: { requestKey, args },
      } = action;

      const argsHash = utils.hashObj(args);

      const {
        [requestKey]: { [argsHash]: invalidData, ...validData },
        ...validRequests
      } = state;

      return {
        [requestKey]: {
          [argsHash]: {
            fetching: true,
          },
          ...validData,
        },
        ...validRequests,
      } as DataRequestsState;
    }
    // TODO @monitz87: handle fetchDataError action
    default: {
      return { ...state };
    }
  }
};
