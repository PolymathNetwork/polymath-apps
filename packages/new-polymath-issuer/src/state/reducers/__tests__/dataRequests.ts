import { reducer } from '../dataRequests';
import * as actions from '~/state/actions/dataRequests';
import { RequestKeys } from '~/types';
import { utils } from '@polymathnetwork/new-shared';

describe('Reducer: dataRequests', () => {
  test('initialState', () => {
    expect(reducer(undefined, {} as any)).toMatchSnapshot();
  });

  const args = {
    foo: 'Foo',
    bar: 'Bar',
  };
  const fetchedIds = ['0', '1', '2'];
  const argsHash = utils.hashObj(args);

  const state = {
    [RequestKeys.GetCheckpointsBySymbol]: {
      [argsHash]: fetchedIds,
    },
    [RequestKeys.GetSecurityTokenBySymbol]: {},
    [RequestKeys.GetDividendsByCheckpoint]: {},
    [RequestKeys.GetErc20DividendsModuleBySymbol]: {},
  };

  test('invalidateRequest invalidates cache for all requests of a certain type', () => {
    const result = reducer(
      state,
      actions.invalidateRequest({
        requestKey: RequestKeys.GetCheckpointsBySymbol,
      })
    );
    expect(result[RequestKeys.GetCheckpointsBySymbol]).toEqual({});
  });

  test('invalidateRequest invalidates cache for a request with a specific set of arguments', () => {
    const result = reducer(
      state,
      actions.invalidateRequest({
        requestKey: RequestKeys.GetCheckpointsBySymbol,
        args,
      })
    );
    expect(
      result[RequestKeys.GetCheckpointsBySymbol][argsHash]
    ).toBeUndefined();
  });

  test('cacheData creates a cache entry', () => {
    const result = reducer(
      undefined,
      actions.cacheData({
        requestKey: RequestKeys.GetCheckpointsBySymbol,
        args,
        fetchedIds,
      })
    );

    expect(result[RequestKeys.GetCheckpointsBySymbol][argsHash]).toEqual(
      fetchedIds
    );
  });
});
