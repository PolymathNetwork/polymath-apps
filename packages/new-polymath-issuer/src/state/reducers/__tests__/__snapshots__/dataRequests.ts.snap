// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer: dataRequests initialState 1`] = `
Object {
  "getCheckpointBySymbolAndId": Object {},
  "getCheckpointsBySymbol": Object {},
  "getDividendBySymbolAndId": Object {},
  "getDividendsByCheckpoint": Object {},
  "getErc20BalanceByAddressAndWallet": Object {},
  "getErc20DividendsModuleBySymbol": Object {},
  "getSecurityTokenBySymbol": Object {},
  "getTaxWithholdingListBySymbol": Object {},
}
`;
